nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/AbstractJsonQueryElasticsearch.java

```diff
@@ -0,0 +1,293 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class AbstractJsonQueryElasticsearch<Q extends JsonQueryParameters> extends AbstractProcessor implements ElasticsearchRestProcessor {
+    public static final Relationship REL_ORIGINAL = new Relationship.Builder().name("original")
+            .description("All original flowfiles that don't cause an error to occur go to this relationship.").build();
+    public static final Relationship REL_HITS = new Relationship.Builder().name("hits")
+            .description("Search hits are routed to this relationship.")
+            .build();
+    public static final Relationship REL_AGGREGATIONS = new Relationship.Builder().name("aggregations")
+            .description("Aggregations are routed to this relationship.")
+            .build();
+
+    public static final AllowableValue SPLIT_UP_YES = new AllowableValue(
+            "splitUp-yes",
+            "Yes",
+            "Split up results."
+    );
+    public static final AllowableValue SPLIT_UP_NO = new AllowableValue(
+            "splitUp-no",
+            "No",
+            "Don't split up results."
+    );
+
+    public static final PropertyDescriptor SPLIT_UP_HITS = new PropertyDescriptor.Builder()
+            .name("el-rest-split-up-hits")
+            .displayName("Split up search results")
+            .description("Split up search results into one flowfile per result.")
+            .allowableValues(SPLIT_UP_NO, SPLIT_UP_YES)
+            .defaultValue(SPLIT_UP_NO.getValue())
+            .required(true)
+            .expressionLanguageSupported(ExpressionLanguageScope.NONE)
+            .build();
+    public static final PropertyDescriptor SPLIT_UP_AGGREGATIONS = new PropertyDescriptor.Builder()
+            .name("el-rest-split-up-aggregations")
+            .displayName("Split up aggregation results")
+            .description("Split up aggregation results into one flowfile per result.")
+            .allowableValues(SPLIT_UP_NO, SPLIT_UP_YES)
+            .defaultValue(SPLIT_UP_NO.getValue())
+            .required(true)
+            .expressionLanguageSupported(ExpressionLanguageScope.NONE)
+            .build();
+
+    private static final Set<Relationship> relationships;
+    private static final List<PropertyDescriptor> propertyDescriptors;
+
+    AtomicReference<ElasticSearchClientService> clientService;
+    String splitUpHits;
+    private String splitUpAggregations;
+
+    final ObjectMapper mapper = new ObjectMapper();
+
+    static {
+        final Set<Relationship> rels = new HashSet<>();
+        rels.add(REL_ORIGINAL);
+        rels.add(REL_FAILURE);
+        rels.add(REL_HITS);
+        rels.add(REL_AGGREGATIONS);
+        relationships = Collections.unmodifiableSet(rels);
+
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
```

Comment:
Probably makes sense for:
* outputting a FlowFile per hit
* outputting a FlowFile per Response (page) when using the Paginated Processors

So will look to do that

Reply:
So one problem with this... the input FlowFile (where it's used in a processor) will need to be transferred to the `original` output with the *first* batch so that the Session can be committed

This *may* be a problem for anyone that's using a Wait/Notify kind of setup - if they want to batch the result outputs but then wait for all hits/pages of results to be processed before continuing with another part of their Flow, they're going to need to know how many hits/pages to expect beforehand and then wait for those *instead* of simply waiting for the `original` FlowFile to appear... is this a problem? Should we only have batching where there's not `input` FlowFile (sounds like a complex processor configuration then)?

Reply:
And, actually, once the input FlowFile has been transferred in the first batch, then it can't be used to create any further result FlowFile for output either

So, basically, I don't think this will work for instances where the processor accepts an incoming FlowFile for the query (unless you know of a way of achieving that) - do we want to retain this functionality at all?

Reply:
Yeah, good call with the `original` issue.  Perhaps if the query was expected to return that many results, you would naturally switch to using `ConsumeElasticsearch`, since successive pages can be run in separate processor runs.  Another good use case for the two separate processors.

Reply:
I'll leave this functionality out of this PR then, if wanted in future, maybe it should be a follow-on ticket?

Reply:
Agreed